### PR TITLE
Add a CURLOPT_SSL_OPTIONS value to disable date checking

### PR DESCRIPTION
**Pull request by TaborKelly**
Add a CURLOPT_SSL_OPTIONS value to disable date checking, but leave all other checks in place. This is particularly useful for embedded devices that lack a real time clock. This is of value because you are still validating the chain of trust. That is, the server certificate is still checked to make sure that there is a chain of trust back to one of the CA certs in your trust store. The only thing that you are ignoring are the dates in the server certificate, and the dates weren't valuable anyway if you don't trust your system time.

### CODE REVIEW

File: docs/libcurl/opts/CURLOPT_SSL_OPTIONS.3 [hunk 1]
```diff
@@ -48,7 +48,7 @@ can't be bypassed.\fP This option may have broader support to accommodate other
 SSL backends in the future.
 https://curl.haxx.se/docs/ssl-compared.html
 
-Added in ?.??.?:
+Added in 7.59.0:
```
Comment (TaborKelly) [inline docs/libcurl/opts/CURLOPT_SSL_OPTIONS.3]: This is actually wrong.

File: docs/libcurl/symbols-in-versions [hunk 2]
```diff
@@ -701,6 +701,7 @@ CURLSSLBACKEND_SCHANNEL         7.34.0
 CURLSSLBACKEND_WOLFSSL          7.49.0
 CURLSSLOPT_ALLOW_BEAST          7.25.0
 CURLSSLOPT_NO_REVOKE            7.44.0
+CURLSSLOPT_NO_DATE_CHECK        7.59.0
```
Comment (TaborKelly) [inline docs/libcurl/symbols-in-versions]: This is actually wrong.

File: lib/vtls/openssl.c [hunk 3]
```diff
@@ -2029,6 +2029,20 @@ set_ssl_version_min_max(long *ctx_options, struct connectdata *conn,
   return CURLE_OK;
 }
 
+static int _verify_no_date_check(int preverify_ok, X509_STORE_CTX *x509_ctx)
+{
+  /* If we have an error related to the validity period, then unset the error.
+     and return success */
+  if(X509_STORE_CTX_get_error(x509_ctx) == X509_V_ERR_CERT_HAS_EXPIRED ||
```
Comment (dscho) [inline lib/vtls/openssl.c]: Does it make sense to check this even if `preverify_ok` is already false? Maybe prepend `preverify_ok &&` to the condition?

File: tests/certs/scripts/genserv.sh [hunk 4]
```diff
@@ -59,6 +59,13 @@ if [ ".$SERIAL" = . ] ; then
 	SERIAL=`/usr/bin/env perl -e "$GETSERIAL"`
 fi
 
+# A hack for the expired certificate test case. We could do better, but it would
+# require using "openssl ca" instead of "openssl x509" for creating the server
+# certificates, as "openssl ca" lets you precisely specify start and end dates.
+if grep -qe "exp$" <<< "$PREFIX"; then
```
Comment (dscho) [inline tests/certs/scripts/genserv.sh]: I think `<<<` is a Bash-ism. Maybe better use `case`, which is available in all POSIX-compliant shells:

```sh
case "$PREFIX" in
*exp)
    DURATION=-$DURATION
    ;;
esac
```

File: tests/libtest/lib2074.c [hunk 5]
```diff
@@ -0,0 +1,67 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) 2013 - 2017, Linus Nielsen Feltzing, <linus@haxx.se>
```
Comment (dscho) [inline tests/libtest/lib2074.c]: Is this based on another source code file? Maybe at least add your name?

**Comment by vszakats** [comment]
Perhaps `CURLSSLOPT_NO_DATE_VERIFY` (vs. current `CURLSSLOPT_NO_DATE_CHECK`) would be a name that aligns more with existing options that fall in the same category as this one (`CURLOPT_SSL_VERIFYPEER`, `CURLOPT_SSL_VERIFYHOST`, `CURLOPT_PROXY_SSL_VERIFYPEER`, `CURLOPT_PROXY_SSL_VERIFYPEER`).

File: lib/vtls/openssl.c (hunk 3 above)
Comment (TaborKelly) [reply lib/vtls/openssl.c]: preverify_ok will be 0 in the case that we care about here. That is, preverify_ok will be 0 if X509_STORE_CTX_get_error(x509_ctx) == X509_V_ERR_CERT_HAS_EXPIRED or X509_STORE_CTX_get_error(x509_ctx) == X509_V_ERR_CERT_NOT_YET_VALID.

Comment (TaborKelly) [reply lib/vtls/openssl.c]: I think that you are correct. We should probably add something like `!preverify_ok && ...` to save the call to  `X509_STORE_CTX_get_error()` if `preverify_ok==1`. I'll work on that tomorrow.

File: lib/vtls/openssl.c [hunk 6]
```diff
@@ -2029,6 +2029,22 @@ set_ssl_version_min_max(long *ctx_options, struct connectdata *conn,
...
 
+static int _verify_no_date_check(int preverify_ok, X509_STORE_CTX *x509_ctx)
+{
+  /* If we have an error related to the validity period, then unset the error.
+     and return success. !preverify_ok is just to save a function call in the
+     normal happy-path case. */
+  if(!preverify_ok &&
+     (X509_STORE_CTX_get_error(x509_ctx) == X509_V_ERR_CERT_HAS_EXPIRED ||
+      X509_STORE_CTX_get_error(x509_ctx) == X509_V_ERR_CERT_NOT_YET_VALID)) {
+     X509_STORE_CTX_set_error(x509_ctx, X509_V_OK); /* Clear "expired" error */
```
Comment (Lekensteyn) [inline lib/vtls/openssl.c]: Indentation is off here by one space

File: tests/certs/scripts/genserv.sh [hunk 7]
```diff
@@ -59,6 +59,15 @@ if [ ".$SERIAL" = . ] ; then
...
 fi
 
+# A hack for the expired certificate test case. We could do better, but it would
+# require using "openssl ca" instead of "openssl x509" for creating the server
+# certificates, as "openssl ca" lets you precisely specify start and end dates.
+case "$PREFIX" in
+*exp)
+    DURATION=-$DURATION
+    ;;
+esac
```
Comment (Lekensteyn) [inline tests/certs/scripts/genserv.sh]: This is very unusual, such certificates should not be encountered in the real world. I can imagine that some implementations refuse such certificates.

Would it be possible to explore the `openssl ca` approach to have more realistic certificates?

**Comment by bagder** [comment]
1. I would like to see an effort to consider @Lekensteyn's remarks for both support in nss + gnutls as well as his point about making sure with a test that the disabling is only for the date and an otherwise  non-matching cert should still be rejected
2. this patch has unfortunately gone conflicting with the current master so it needs a rebase/fix

**Comment by TaborKelly** [comment]
Okay, I've moved onto a new role at a new company and no longer need this support. You can let it die.

**Comment by danielgustafsson** [comment]
Unless someone is interested in adopting this code I propose we close this.

**Comment by danielgustafsson** [comment]
Closed due to inactivity. Thanks for submitting this to the project, if you ever wish to return to this patch then feel free to reopen.

**Pull request closed without merging**